audio/data.py

```diff
@@ -260,64 +288,75 @@ def __init__(self, items, path, config=AudioConfig(), **kwargs):
-                                does not match config sample rate {cfg._sr} 
+    def open(self, fn): # file name, it seems
+        fn=Path(fn)
+        if self.path is not None and not fn.exists() and str(self.path) not in str(fn): fn = self.path/item
+        if self.config.use_spectro:
+            item=self.add_spectro(fn)
+        else:
+            if self.config.max_to_pad or self.config.segment_size:
+                pad_len = self.config.max_to_pad if self.config.max_to_pad is not None else self.config.segment_size
+                func_to_add = lambda s1,s2: tfm_padtrim_signal(s1, int(pad_len/1000*item.s1), pad_mode="zeros")
+            item=open_audio(fn, func_to_add)
+            self._validate_consistencies(item)
+        return item
+
+    def _validate_consistencies(self, item):
+        if(self.config._sr is not None and item.sr != self.config._sr):
+            raise ValueError(f'''Multiple sample rates detected. Sample rate {item.sr} of file {item.path} 
+                                does not match config sample rate {self.config._sr} 
                                 this means your dataset has multiple different sample rates, 
                                 please choose one and set resample_to to that value''')
-        if(sig.shape[0] > 1):
-            if not cfg.downmix:
-                warnings.warn(f'''Audio file {p} has {sig.shape[0]} channels, automatically downmixing to mono, 
-                                set AudioConfig.downmix=True to remove warnings''')
-            sig = DownmixMono(channels_first=True)(sig)
-        if cfg.max_to_pad or cfg.segment_size:
-            pad_len = cfg.max_to_pad if cfg.max_to_pad is not None else cfg.segment_size
-            sig = tfm_padtrim_signal(sig, int(pad_len/1000*sr), pad_mode="zeros")
-
-        mel = None
-        if cfg.use_spectro:
-            if cfg.mfcc: mel = MFCC(sr=sr, n_mfcc=cfg.sg_cfg.n_mfcc, melkwargs=cfg.sg_cfg.mel_args())(sig)
-            else:
-                mel = MelSpectrogram(**(cfg.sg_cfg.mel_args()))(sig)
-                if cfg.sg_cfg.to_db_scale: mel = SpectrogramToDB(top_db=cfg.sg_cfg.top_db)(mel)
-            mel = mel.permute(0, 2, 1)
-            if cfg.standardize: mel = standardize(mel)
-            if cfg.delta: mel = torch.stack([m.squeeze(0) for m in [mel, torchdelta(mel), torchdelta(mel, order=2)]]) 
-            if cfg.cache:
-                os.makedirs(image_path.parent, exist_ok=True)
-                torch.save(mel, image_path)
-                _record_cache_contents(cfg, [image_path])
-            start, end = None, None
-            if cfg.duration and cfg._processed: 
-                mel, start, end = tfm_crop_time(mel, cfg._sr, cfg.duration, cfg.sg_cfg.hop, cfg.pad_mode)
-        return AudioItem(sig=sig.squeeze(), sr=sr, spectro=mel, path=item, start=start, end=end)
+        if(self.config._nchannels != item.nchannels):
```

Comment:
Need to have a check here that `self.config._nchannels` isn't None.

We want users to be able to generate AudioLists with many sample rates or channels, it's only at train time that this is a problem. Preprocessing to ensure single SR and n_channels is done at the same time as labeling, that's why there are lines to set `config._sr` and `config._nchannels` to be their correct values. Before that point they are both `None` and if they are `None` that means we don't want to validate.